// @arbitrum/sdk-nitro/src/lib/abi/factories/IRollupEventBridge__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import { Provider } from "@ethersproject/providers";
import type {
  IRollupEventBridge,
  IRollupEventBridgeInterface,
} from "../IRollupEventBridge";

const _abi = [
  {
    inputs: [],
    name: "bridge",
    outputs: [
      {
        internalType: "contract IBridge",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "_bridge",
        type: "address",
      },
      {
        internalType: "address",
        name: "_rollup",
        type: "address",
      },
    ],
    name: "initialize",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "rollup",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "chainId",
        type: "uint256",
      },
    ],
    name: "rollupInitialized",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class IRollupEventBridge__factory {
  static readonly abi = _abi;
  static createInterface(): IRollupEventBridgeInterface {
    return new utils.Interface(_abi) as IRollupEventBridgeInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IRollupEventBridge {
    return new Contract(address, _abi, signerOrProvider) as IRollupEventBridge;
  }
}
